File: ac_diskstructs/heap.py
import ac_diskstructs.impl as LIB


class HeapIterator(object):
    def __init__(self, ptr):
        self.__ptr = ptr

    def delete(self):
        LIB.HeapIteratorDelete(self.__ptr)

        self.delete = lambda: None

    def __del__(self):
        self.delete()

    def __iter__(self):
        return self

    def __next__(self):
        item = LIB.HeapIteratorNext(self.__ptr)

        if item is None:
            self.delete()
            raise StopIteration()

        return item


class Heap(object):
    def __init__(self, path, *, record_size=None):
        self.__ptr = (LIB.HeapOpen(path) if record_size is None else LIB.HeapNew(path, record_size))

    def record_size(self):
        return LIB.HeapRecordSize(self.__ptr)

    # Free memory
    def delete(self):
        LIB.HeapDelete(self.__ptr)

        self.delete = lambda: None

    def __del__(self):
        self.delete()

    # Finish write operations
    def close(self):
        LIB.HeapClose(self.__ptr)

        self.close = lambda: None

    def __getitem__(self, prefix):
        it = self.get_all(prefix)

        for rv in it:
            it.delete()
            return rv

        raise KeyError(prefix)

    def get(self, prefix, default=None):
        try:
            return self[prefix]

        except KeyError:
            return default

    def insert(self, value):
        LIB.HeapInsert(self.__ptr, value)

    def get_all(self, prefix):
        return HeapIterator(LIB.HeapGetAll(self.__ptr, prefix))

File: src/heap.cpp
#include "heap.hpp"

#include <ac-library/containers/persistent/binary_heap/binary_heap.hpp>
#include <cstdint>

namespace NACPyDiskStructs {
    using namespace NAC;

    PyObject* HeapOpen(PyObject*, PyObject* args) {
        const char* path;

        if (!PyArg_ParseTuple(args, "s", &path)) {
            return nullptr;
        }

        return Py_BuildValue("K", (uintptr_t)new TPersistentBinaryHeap(path));
    }

    PyObject* HeapNew(PyObject*, PyObject* args) {
        const char* path;
        unsigned long long int recordSize;

        if (!PyArg_ParseTuple(args, "sK", &path, &recordSize)) {
            return nullptr;
        }

        return Py_BuildValue("K", (uintptr_t)new TPersistentBinaryHeap(path, recordSize));
    }

    PyObject* HeapDelete(PyObject*, PyObject* args) {
        unsigned long long int ptr;

        if (!PyArg_ParseTuple(args, "K", &ptr)) {
            return nullptr;
        }

        delete (TPersistentBinaryHeap*)ptr;

        Py_RETURN_NONE;
    }

    PyObject* HeapInsert(PyObject*, PyObject* args) {
        unsigned long long int ptr;
        const char* value;
        Py_ssize_t valueLength;

        if (!PyArg_ParseTuple(args, "Ks#", &ptr, &value, &valueLength)) {
            return nullptr;
        }

        ((TPersistentBinaryHeap*)ptr)->Insert(value);

        Py_RETURN_NONE;
    }

    PyObject* HeapClose(PyObject*, PyObject* args) {
        unsigned long long int ptr;

        if (!PyArg_ParseTuple(args, "K", &ptr)) {
            return nullptr;
        }

        ((TPersistentBinaryHeap*)ptr)->Close();

        Py_RETURN_NONE;
    }

    PyObject* HeapRecordSize(PyObject*, PyObject* args) {
        unsigned long long int ptr;

        if (!PyArg_ParseTuple(args, "K", &ptr)) {
            return nullptr;
        }

        return Py_BuildValue("K", ((TPersistentBinaryHeap*)ptr)->RecordSize());
    }

    PyObject* HeapGetAll(PyObject*, PyObject* args) {
        unsigned long long int ptr;
        const char* prefix;
        Py_ssize_t prefixLength;

        if (!PyArg_ParseTuple(args, "Ks#", &ptr, &prefix, &prefixLength)) {
            return nullptr;
        }

        return Py_BuildValue("K", (uintptr_t)new TPersistentBinaryHeap::TIterator(((TPersistentBinaryHeap*)ptr)->GetAll(TBlob(prefixLength, prefix))));
    }

    PyObject* HeapIteratorDelete(PyObject*, PyObject* args) {
        unsigned long long int ptr;

        if (!PyArg_ParseTuple(args, "K", &ptr)) {
            return nullptr;
        }

        delete (TPersistentBinaryHeap::TIterator*)ptr;

        Py_RETURN_NONE;
    }

    PyObject* HeapIteratorNext(PyObject*, PyObject* args) {
        unsigned long long int ptr;

        if (!PyArg_ParseTuple(args, "K", &ptr)) {
            return nullptr;
        }

        TBlob data;

        if (((TPersistentBinaryHeap::TIterator*)ptr)->Next(data)) {
            return Py_BuildValue("y#", data.Data(), data.Size());
        }

        Py_RETURN_NONE;
    }
}
